### PR TITLE
Extensibility changes for MarkStep

### PR DESCRIPTION
**Pull request by erozenfeld**
Make several methods protected and add an extensibility point for MarkStep to allow derived classes to customize the behavior.

### CODE REVIEW

**Comment by erozenfeld** [comment]
@marek-safar PTAL

File: linker/Mono.Linker.Steps/MarkStep.cs [hunk 1]
```diff
@@ -860,7 +866,9 @@ static bool IsDefaultConstructor (MethodDefinition method)
 			return IsConstructor (method) && !method.HasParameters;
 		}
 
-		static bool IsConstructor (MethodDefinition method)
+		protected static MethodPredicate IsConstructorPredicate = new MethodPredicate (IsConstructor);
+
+		protected static bool IsConstructor (MethodDefinition method)
```
Comment (marek-safar) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Better to name it correctly, e.g `IsInstanceConstructor` before exposing it

File: linker/Mono.Linker.Steps/MarkStep.cs [hunk 2]
```diff
@@ -860,7 +866,9 @@ static bool IsDefaultConstructor (MethodDefinition method)
 			return IsConstructor (method) && !method.HasParameters;
 		}
 
-		static bool IsConstructor (MethodDefinition method)
+		protected static MethodPredicate IsConstructorPredicate = new MethodPredicate (IsConstructor);
```
Comment (marek-safar) [inline linker/Mono.Linker.Steps/MarkStep.cs]: This seems redundant, compiler can do good job if you just use MethodGroupExpression for predicates

Comment (erozenfeld) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Removed.

File: linker/Mono.Linker.Steps/MarkStep.cs (hunk 1 above)
Comment (erozenfeld) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Done.

**Comment by erozenfeld** [comment]
@marek-safar I addressed your feedback. PTAL.

**Comment by marek-safar** [comment]
@erozenfeld could you rebase your changes

**Comment by erozenfeld** [comment]
@marek-safar Done.

**Pull request merged**